lib/tpm2_options.c

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright (c) 2016, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of Intel Corporation nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <getopt.h>
+#include <unistd.h>
+
+#include "log.h"
+#include "tpm2_options.h"
+#include "tpm2_util.h"
+
+#ifdef HAVE_TCTI_DEV
+#include "tpm2_tools_tcti_device.h"
+#endif
+#ifdef HAVE_TCTI_SOCK
+#include "tpm2_tools_tcti_socket.h"
+#endif
+#ifdef HAVE_TCTI_TABRMD
+#include "tpm2_tools_tcti_abrmd.h"
+#endif
+
+/*
+ * Default TCTI: this is a bit awkward since we allow users to enable /
+ * disable TCTIs using ./configure --with/--without magic.
+ * As simply put as possible:
+ * if the tabrmd TCTI is enabled, it's the default.
+ * else if the socket TCTI is enabled it's the default.
+ * else if the device TCTI is enabled it's the default.
+ * We do this to preserve the current default / expected behavior (use of
+ * the socket TCTI).
+ */
+#ifdef HAVE_TCTI_TABRMD
+  #define TCTI_DEFAULT_STR  "abrmd"
+#elif HAVE_TCTI_SOCK
+  #define TCTI_DEFAULT_STR  "socket"
+#elif  HAVE_TCTI_DEV
+  #define TCTI_DEFAULT_STR  "device"
+#endif
+
+#ifndef VERSION
+  #warning "VERSION Not known at compile time, not embedding..."
+  #define VERSION "UNKNOWN"
+#endif
+
+#define TPM2TOOLS_ENV_TCTI_NAME      "TPM2TOOLS_TCTI_NAME"
+
+struct tpm2_options {
+    tpm2_option_handler on_opt;
+    tpm2_arg_handler on_arg;
```

Comment:
I suggested calling these on_callback or on_handler, I really think is much clear that on_opt.

Reply:
Sure, not a problem. I think I lost some of your comments in the noise.